### DISCUSSION TITLE
Disable source code caching

### DISCUSSION THREAD

**Discussion by AndMrzv** [netbox-community/netbox#9429] [Q&A] [answered]
TL;DR: how to disable every possible caching mechanism for netbox?

I write custom modules and often netbox shows me output of what was in the code few versions of file ago, this slows down development. The problem is most noticeable with validators in /opt/netbox/netbox/ directory, hadn't had much of a trouble with reports in /etc/netbox/. Tried with no success so far:
- enabling DEBUG in both netbox and django configuration files;
- changing django caching to django.core.cache.backends.dummy.DummyCache;
- setting CACHE_MIDDLEWARE_SECONDS to zero, adding and setting CACHE_TIMEOUT (turns out to be deprecated) to zero too;
- restarting dozen of times.

Emptying out redis cache settings breaks netbox. Deleting and rebuilding all the containers and volumes on every edit is not optimal.
What am I missing?

**Comment by jeremystretch** [comment, marked as answer]
Are you running NetBox's using [Django's development server](https://docs.djangoproject.com/en/4.0/ref/django-admin/#runserver)? This will automatically restart when changes to the code base are detected.

**Reply by AndMrzv** [reply]
I did "python3 manage.py runserver" right now and it said
>Starting development server at http://127.0.0.1:8000/

so I guess so (it did show the latest output on the first query after this command but on the second run it jumped 4 versions back). If that's not enough how do I check?

**Reply by kkthxbye-code** [reply]
> Deleting and rebuilding all the containers and volumes on every edit

The docker containers are not intended for development.

**Reply by AndMrzv** [reply]
>The docker containers are not intended for development.

They are not intended for stateful apps as well yet here we are.

**Reply by kkthxbye-code** [reply]
I think you are misunderstanding something fundamental about how the images function.

**Reply by kkthxbye-code** [reply]
To make sure you understand, there is no out of image code caching going on at all. You mention validators, which I assume you mount to the image. These will not be reloaded on demand, I feel like you expect it to function like dropping php files in a web root. The code is however picked up when you start a new container with the image.

The reason you are seeing it flip between code versions is that the default docker image can start up to 4 nginx-unit processes. What version of the code you are getting depends on what process handles the request, because the processes are started on demand and closed when not used. If you hit one of the long running ones, you will get old code. The python code is not loaded every request, it's only loaded when the application server starts a new process.

I can recommend reading up on how containers function in general and how django applications are served via. wsgi.

If you want to develop in the docker image you should modify it to run the development server as Jeremy mentioned, instead of running it with nginx-unit.

**Reply by AndMrzv** [reply]
Sure, not trying to flex here, feel free to correct any misunderstanding.
I got directories attached as volumes to the container (image is a different thing, static), when I edit something I can attach to the container and see that the file is in fact changed without extra actions. But okay, the app server may not reload them on each request (indeed, I have not expected this having done a lot of php editing on live servers back in the days - disabling memcached/opcache/whatever was enough).
Later I've noticed that web port was mapped to 8080, querying directly django dev server on port 8000 gives me "Static Media Failure" page. /opt/netbox/netbox/static/ folder with all the needed files has the same owner as other netbox files.
The image is netboxcommunity/netbox:v3.1.9 by the way, I see @kkthxbye-code being active there too so maybe you'll help me figure out what exactly do I have to modify to run it properly (without fully ditching the environment set up by ops and switching to running my own install)?

**Reply by kkthxbye-code** [reply]
> querying directly django dev server on port 8000 gives me "Static Media Failure"

`python manage.py runserver --insecure`

Alternatively set `DEBUG = True` in configuration.py

**Reply by AndMrzv** [reply]
It did the trick. I see django server restarting on every file save, tested with a number of edits and always got the latest output.
Big thanks!
